// gatsby-plugin-use-dark-mode/theme-hydration-script-tag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _terser = _interopRequireDefault(require("terser"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// Adapted from:
// https://github.com/donavon/use-dark-mode/blob/develop/noflash.js.txt
function generateNoFlashScript({
  classNameDark,
  classNameLight,
  storageKey
}) {
  return `
    (function(classNameDark, classNameLight, storageKey) {
      function setClassOnDocumentBody(darkMode) {
        document.body.classList.add(darkMode ? classNameDark : classNameLight);
        document.body.classList.remove(darkMode ? classNameLight : classNameDark);
      }

      var preferDarkQuery = '(prefers-color-scheme: dark)';
      var mql = window.matchMedia(preferDarkQuery);
      var supportsColorSchemeQuery = mql.media === preferDarkQuery;
      var localStorageTheme = null;
      try {
        localStorageTheme = localStorage.getItem(storageKey);
      } catch (err) {}
      var localStorageExists = localStorageTheme !== null;
      if (localStorageExists) {
        localStorageTheme = JSON.parse(localStorageTheme);
      }

      // Determine the source of truth
      if (localStorageExists) {
        // source of truth from localStorage
        setClassOnDocumentBody(localStorageTheme);
      } else if (supportsColorSchemeQuery) {
        // source of truth from system
        setClassOnDocumentBody(mql.matches);
        localStorage.setItem(storageKey, mql.matches);
      } else {
        // source of truth from document.body
        var isDarkMode = document.body.classList.contains(classNameDark);
        localStorage.setItem(storageKey, JSON.stringify(isDarkMode));
      }
    })('${classNameDark}', '${classNameLight}', '${storageKey}');
  `;
}

function ThemeHydrationScriptTag({
  classNameDark,
  classNameLight,
  storageKey,
  minify
}) {
  const noFlashScript = generateNoFlashScript({
    classNameDark,
    classNameLight,
    storageKey
  });
  const finalNoFlashScript = minify ? _terser.default.minify(noFlashScript).code || '' : noFlashScript; // eslint-disable-next-line react/no-danger

  return /*#__PURE__*/React.createElement("script", {
    dangerouslySetInnerHTML: {
      __html: finalNoFlashScript
    }
  });
}

ThemeHydrationScriptTag.propTypes = {
  classNameDark: _propTypes.default.string,
  classNameLight: _propTypes.default.string,
  storageKey: _propTypes.default.string,
  minify: _propTypes.default.bool
};
ThemeHydrationScriptTag.defaultProps = {
  classNameDark: 'dark-mode',
  classNameLight: 'light-mode',
  storageKey: 'darkMode',
  minify: true
};
var _default = ThemeHydrationScriptTag;
exports.default = _default;
