// gatsby-plugin-typegen/services/emitSchema.js
"use strict";

exports.__esModule = true;
exports.makeEmitSchemaService = void 0;

var _graphql = require("gatsby/graphql");

var _utils = require("../internal/utils");

const makeEmitSchemaService = ({
  configMap,
  reporter,
  writeFileContent
}) => {
  return async schema => {
    const entries = Object.entries(configMap);

    if (entries.length === 0) {
      return;
    }

    await Promise.all(entries.map(async ([filePath, config]) => {
      const activity = reporter.activity(`emit schema into ${filePath}`);
      activity.start();
      const {
        format,
        commentDescriptions,
        omitPluginMetadata
      } = config;

      try {
        let outputSchema = schema;

        if (omitPluginMetadata) {
          outputSchema = (0, _utils.filterPluginSchema)(outputSchema);
        }

        switch (format) {
          case 'sdl':
            {
              await writeFileContent(filePath, (0, _graphql.printSchema)(outputSchema, {
                commentDescriptions
              }));
              break;
            }

          case 'introspection':
            {
              await writeFileContent(filePath, JSON.stringify((0, _graphql.introspectionFromSchema)(outputSchema, {
                descriptions: true,
                schemaDescription: true,
                directiveIsRepeatable: true,
                inputValueDeprecation: true,
                specifiedByUrl: true
              }), null, 2));
              break;
            }
        }
      } catch (e) {
        activity.panic(e);
      } finally {
        activity.end();
      }
    }));
  };
};

exports.makeEmitSchemaService = makeEmitSchemaService;
