// gatsby-plugin-typegen/internal/config.js
"use strict";

exports.__esModule = true;
exports.validateConfig = void 0;

var path = _interopRequireWildcard(require("path"));

var _utils = require("./utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const defaultSchemaOutputOption = Object.freeze({
  format: 'introspection',
  commentDescriptions: true,
  omitPluginMetadata: true
});
const defaultDocumentOutputOption = Object.freeze({
  format: 'graphql'
});

const validateConfig = ({
  options,
  basePath,
  reporter
}) => {
  // There are no required properties (yet), so must be compatible.
  const pluginOptions = options;
  const {
    language = 'typescript',
    namespace = 'GatsbyTypes',
    includeResolvers = false,
    autoFix,
    autofix,
    emitSchema: emitSchemaOptionMap = {},
    emitPluginDocuments: legacyEmitPluginDocumentsOptionMap = {},
    emitPluginDocument: emitPluginDocumentsOptionMap = {},
    scalars = {}
  } = pluginOptions;

  if (autoFix != null) {
    reporter.warn('The `autoFix` option is deprecated. Use `autofix` instead.');
  }

  if (includeResolvers) {
    reporter.warn('The `includeResolvers` option is deprecated. It will be removed in v4');
  }

  const emitSchema = {};

  for (const [key, options] of Object.entries(emitSchemaOptionMap)) {
    if (!options) continue;

    if (options === true) {
      emitSchema[key] = { ...defaultSchemaOutputOption,
        // Infer format option based on filename.
        format: /\.(graphql|graphqls|gql)$/.test(key) ? 'sdl' : 'introspection'
      };
    } else {
      emitSchema[key] = { ...defaultSchemaOutputOption,
        ...options
      };
    }
  }

  const emitPluginDocument = {}; // TBD: remove next major release

  for (const [key, options] of Object.entries(legacyEmitPluginDocumentsOptionMap)) {
    if (!options) continue;

    if (options === true) {
      emitPluginDocument[key] = { ...defaultDocumentOutputOption,
        format: /\.json$/.test(key) ? 'json' : 'graphql'
      };
    } else {
      emitPluginDocument[key] = { ...defaultDocumentOutputOption,
        ...options
      };
    }
  }

  for (const [key, options] of Object.entries(emitPluginDocumentsOptionMap)) {
    if (!options) continue;

    if (options === true) {
      emitPluginDocument[key] = { ...defaultDocumentOutputOption,
        // Infer format option based on filename.
        format: /\.json$/.test(key) ? 'json' : 'graphql'
      };
    } else {
      emitPluginDocument[key] = { ...defaultDocumentOutputOption,
        ...options
      };
    }
  }

  const outputPath = pluginOptions.outputPath || (language === 'typescript' ? path.resolve(basePath, 'src/__generated__/gatsby-types.d.ts') : path.resolve(basePath, 'src/__generated__/gatsby-types.js.flow'));

  if (language === 'typescript' !== /\.tsx?$/.test(outputPath)) {
    reporter.warn(reporter.stripIndent`
        The language you specified is not match to file extension.
          - language: ${(0, _utils.formatLanguage)(language)}
          - outputPath: ${outputPath}
      `);
  }

  for (const type of _utils.gatsbyInternalScalars) {
    if (scalars[type]) {
      reporter.warn(`You couldn't override type for \`${type}\` scalar because it is reserved by Gatsby internal.`);
      delete scalars[type];
    }
  }

  return Object.freeze({
    language,
    namespace,
    outputPath,
    includeResolvers,
    autofix: autofix === true || autoFix === true || autofix !== false,
    emitSchema,
    emitPluginDocument,
    scalars
  });
};

exports.validateConfig = validateConfig;
