// gatsby-plugin-typegen/internal/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.definitionIsEqual = definitionIsEqual;
exports.filterDevOnlySchema = filterDevOnlySchema;
exports.filterPluginSchema = filterPluginSchema;
exports.gatsbyInternalScalars = exports.formatLanguage = void 0;
exports.isFragmentDefinition = isFragmentDefinition;
exports.isQueryDefinition = isQueryDefinition;
exports.isTargetDefinition = isTargetDefinition;
exports.isThirdpartyFragment = isThirdpartyFragment;
exports.sortDefinitions = sortDefinitions;
exports.stabilizeSchema = stabilizeSchema;

var _slugify = _interopRequireDefault(require("slugify"));

var _lodash = _interopRequireDefault(require("lodash"));

var _graphql = require("gatsby/graphql");

var _utils = require("@graphql-tools/utils");

const formatLanguage = lang => lang === 'typescript' ? 'TypeScript' : 'Flow';

exports.formatLanguage = formatLanguage;

function definitionIsEqual(a, b) {
  return a.name === b.name && a.hash === b.hash;
}

function isFragmentDefinition(def) {
  return def.isFragment;
}

function isQueryDefinition(def) {
  return !def.isFragment;
}
/**
 * return `true` if the given definition is assumed to be generated from unnamed query.
 */


function guessIfUnnnamedQuery({
  isStaticQuery,
  name,
  filePath
}) {
  const queryType = isStaticQuery ? 'static' : 'page'; // See https://github.com/gatsbyjs/gatsby/blob/d163724/packages/gatsby/src/query/file-parser.js#L33

  const generatedQueryName = (0, _slugify.default)(filePath, {
    replacement: ' ',
    lower: false
  });

  const pattern = _lodash.default.camelCase(`${queryType}-${generatedQueryName}`);

  return name.startsWith(pattern);
}
/**
 * return `true` if the given definition is assumed to be included by third-party plugins or gatsby internal.
 */


function guessIfThirdpartyDefinition({
  filePath
}) {
  return /(node_modules|\.yarn|\.cache)/.test(filePath);
}

function isThirdpartyFragment(def) {
  return isFragmentDefinition(def) && guessIfThirdpartyDefinition(def);
}

function isTargetDefinition(def) {
  return !(guessIfThirdpartyDefinition(def) || guessIfUnnnamedQuery(def));
} // from https://github.com/gatsbyjs/gatsby/blob/6365768/packages/gatsby/src/schema/print.js#L33-L48


const gatsbyInternalScalars = ['Boolean', 'Buffer', 'Date', 'Float', 'ID', 'Int', 'Internal', 'InternalInput', 'JSON', 'Json', 'Node', 'NodeInput', 'Query', 'String'];
exports.gatsbyInternalScalars = gatsbyInternalScalars;

const fieldFilterFromCoordinates = coords => coord => {
  const target = [coord.typeName, coord.fieldName, coord.argName].filter(Boolean).join('.');
  return !coords.includes(target);
};
/**
 * Remove fields and args used only at development time from schema output.
 */


function filterDevOnlySchema(schema) {
  return (0, _utils.mapSchema)((0, _utils.filterSchema)({
    schema,
    fieldFilter: (typeName, fieldName) => fieldFilterFromCoordinates(['Site.host', 'Site.port', 'SiteFilterInput.host', 'SiteFilterInput.port'])({
      typeName,
      fieldName
    }),
    argumentFilter: (typeName, fieldName, argName) => fieldFilterFromCoordinates(['Query.site.host', 'Query.site.port'])({
      typeName,
      fieldName,
      argName
    })
  }), {
    [_utils.MapperKind.ENUM_TYPE]: type => {
      if (type.name === 'SiteFieldsEnum') {
        const config = type.toConfig();
        delete config.values['host'];
        delete config.values['port'];
        return new _graphql.GraphQLEnumType(config);
      }

      return undefined;
    }
  });
}
/**
 * Remove the plugin options schema.
 * Almost all users do not use it, but it unnecessarily increases the schema output size.
 */


function filterPluginSchema(schema) {
  return (0, _utils.mapSchema)((0, _utils.filterSchema)({
    schema,
    typeFilter: typeName => !typeName.startsWith('SitePlugin'),
    fieldFilter: (typeName, fieldName) => fieldFilterFromCoordinates(['Query.allSitePlugin', 'Query.sitePlugin', 'SitePage.pluginCreatorId'])({
      typeName,
      fieldName
    }),
    inputObjectFieldFilter: (typeName, fieldName) => fieldFilterFromCoordinates(['SitePageFilterInput.pluginCreator', 'SitePageFilterInput.pluginCreatorId'])({
      typeName,
      fieldName
    }),
    argumentFilter: (typeName, fieldName, argName) => fieldFilterFromCoordinates(['Query.sitePage.pluginCreator', 'Query.sitePage.pluginCreatorId'])({
      typeName,
      fieldName,
      argName
    })
  }), {
    [_utils.MapperKind.ENUM_TYPE]: type => {
      if (type.name === 'SitePageFieldsEnum') {
        const config = type.toConfig();

        for (const key of Object.keys(config.values)) {
          if (key.startsWith('pluginCreator')) {
            delete config.values[key];
          }
        }

        return new _graphql.GraphQLEnumType(config);
      }

      return undefined;
    }
  });
}

function stabilizeSchema(schema) {
  return (0, _graphql.lexicographicSortSchema)(filterDevOnlySchema(schema));
}

function sortDefinitions(a, b) {
  const aKey = a.name;
  const bKey = b.name;

  if (aKey < bKey) {
    return -1;
  }

  if (aKey > bKey) {
    return 1;
  }

  return 0;
}
