// gatsby-plugin-typegen/gatsby-node.js
"use strict";

exports.__esModule = true;
exports.pluginOptionsSchema = exports.onPreBootstrap = exports.onCreateDevServer = void 0;

var _xstate = require("xstate");

var _graphql = require("gatsby/graphql");

var _config = require("./internal/config");

var _machine = require("./internal/machine");

var _utils = require("./internal/utils");

var _emitSchema = require("./services/emitSchema");

var _emitPluginDocument = require("./services/emitPluginDocument");

var _autofix = require("./services/autofix");

var _codegen = require("./services/codegen");

var _utils2 = require("./utils");

const pluginOptionsSchema = ({
  Joi
}) => {
  const documentOutputOptionsSchema = Joi.object({
    format: Joi.string().valid('introspection', 'sdl').default('sdl').required(),
    commentDescriptions: Joi.boolean().default(true)
  }).required();
  return Joi.object({
    language: Joi.string().valid('typescript', 'flow').default('typescript'),
    namespace: Joi.string().default('GatsbyTypes'),
    outputPath: Joi.string().default('src/__generated__/gatsby-types.d.ts'),
    includeResolvers: Joi.boolean().default(false),
    autoFix: Joi.boolean(),
    autofix: Joi.boolean().default(true),
    scalars: Joi.object().pattern(/\w[\w\d\-_]+/, Joi.string().required()).default({}),
    emitSchema: Joi.object().pattern(/.+/, [Joi.boolean().required(), documentOutputOptionsSchema]),
    emitPluginDocuments: Joi.object().pattern(/.+/, [Joi.boolean().required(), documentOutputOptionsSchema]),
    emitPluginDocument: Joi.object().pattern(/.+/, [Joi.boolean().required(), documentOutputOptionsSchema])
  });
}; // eslint-disable-next-line @typescript-eslint/no-explicit-any


exports.pluginOptionsSchema = pluginOptionsSchema;
let spawnedMachine = null;

const onPreBootstrap = ({
  store,
  emitter,
  reporter
}, options) => {
  const typegenReporter = makeTypegenReporter(reporter);
  const pluginOptions = options;
  const {
    program
  } = store.getState();
  const basePath = program.directory;
  const config = (0, _config.validateConfig)({
    basePath,
    options: pluginOptions,
    reporter: typegenReporter
  });
  typegenReporter.verbose(reporter.stripIndent`
    loaded configuration
    ${JSON.stringify(pluginOptions, null, 2)}
  `);

  if ((0, _utils2.isCloudBuild)(process.env)) {
    typegenReporter.verbose(reporter.stripIndent`
      skip running on cloud build
    `);
    return;
  }

  const emitSchema = (0, _emitSchema.makeEmitSchemaService)({
    configMap: config.emitSchema,
    reporter: typegenReporter,
    writeFileContent: _utils2.writeFileContent
  });
  const emitPluginDocument = (0, _emitPluginDocument.makeEmitPluginDocumentService)({
    configMap: config.emitPluginDocument,
    reporter: typegenReporter,
    writeFileContent: _utils2.writeFileContent
  });
  const autofix = (0, _autofix.makeAutofixService)({ ...config,
    readFileContent: _utils2.readFileContent,
    writeFileContent: _utils2.writeFileContent,
    reporter: typegenReporter
  });
  const codegen = (0, _codegen.makeCodegenService)({ ...config,
    customScalars: config.scalars,
    reporter: typegenReporter,
    writeFileContent: _utils2.writeFileContent
  });
  const typegenService = (0, _xstate.interpret)(_machine.typegenMachine.withContext({
    config,
    debouncingDelay: 500,
    devMode: false,
    reporter: typegenReporter
  }).withConfig({
    actions: {
      reportEmitSchemaError: (context, event) => {
        context.reporter.error('error occurred while running emitSchema service', event.data);
      },
      reportEmitPluginDocumentError: (context, event) => {
        context.reporter.error('error occurred while running emitPluginDocument service', event.data);
      },
      reportCodegenError: (context, event) => {
        context.reporter.error('error occurred while running codegen service', event.data);
      },
      reportAutofixError: (context, event) => {
        context.reporter.error('error occurred while running autofix service', event.data);
      }
    },
    services: {
      emitSchema: context => {
        if (!context.schema) {
          return Promise.reject(new Error('schema is not initilaized'));
        }

        return emitSchema(context.schema);
      },
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore for typegen bug
      emitPluginDocument: context => emitPluginDocument(context.thirdpartyFragments || []),
      codegen: context => {
        var _context$trackedDefin;

        if (!context.schema) {
          return Promise.reject(new Error('schema is not initilaized'));
        }

        return codegen({
          schema: context.schema,
          documents: [...(((_context$trackedDefin = context.trackedDefinitions) === null || _context$trackedDefin === void 0 ? void 0 : _context$trackedDefin.values()) || [])].sort(_utils.sortDefinitions).map(definitionMeta => ({
            document: {
              kind: _graphql.Kind.DOCUMENT,
              definitions: [definitionMeta.def]
            },
            hash: definitionMeta.hash.toString()
          }))
        });
      },
      autofix: (context, event) => {
        if (!context.config.autofix) {
          var _event$files, _context$trackedDefin2;

          const files = (_event$files = event.files) !== null && _event$files !== void 0 ? _event$files : [...(((_context$trackedDefin2 = context.trackedDefinitions) === null || _context$trackedDefin2 === void 0 ? void 0 : _context$trackedDefin2.values()) || [])].map(def => def.filePath);
          return autofix(files);
        }

        return Promise.resolve();
      }
    }
  }));
  typegenService.start();
  typegenService.onTransition(({
    event,
    value,
    changed
  }) => {
    typegenReporter.verbose(`on ${event.type}`);

    if (changed) {
      typegenReporter.verbose(` ⤷ transition to ${JSON.stringify(value)}`);
    } else {
      typegenReporter.verbose(' ⤷ skipped');
    }
  });
  spawnedMachine = typegenService;
  emitter.on('SET_SCHEMA', () => {
    typegenService.send({
      type: 'SET_SCHEMA',
      schema: store.getState().schema
    });
  });
  emitter.on('SET_GRAPHQL_DEFINITIONS', () => {
    typegenService.send({
      type: 'SET_GRAPHQL_DEFINITIONS',
      definitions: store.getState().definitions
    });
  });
};

exports.onPreBootstrap = onPreBootstrap;

const onCreateDevServer = () => {
  var _spawnedMachine;

  (_spawnedMachine = spawnedMachine) === null || _spawnedMachine === void 0 ? void 0 : _spawnedMachine.send('CREATE_DEV_SERVER');
};

exports.onCreateDevServer = onCreateDevServer;

function makeTypegenReporter(reporter) {
  const formatMessage = message => `[typegen] ${message}`;

  return {
    stripIndent: reporter.stripIndent,
    log: message => reporter.log(formatMessage(message)),
    info: message => reporter.info(formatMessage(message)),
    warn: message => reporter.warn(formatMessage(message)),
    verbose: message => reporter.verbose(formatMessage(message)),
    error: (message, e) => reporter.error(formatMessage(message), e, 'gatsby-plugin-typegen'),
    panic: (message, e) => reporter.panic(formatMessage(message), e, 'gatsby-plugin-typegen'),
    panicOnBuild: (message, e) => reporter.panicOnBuild(formatMessage(message), e, 'gatsby-plugin-typegen'),
    activity: message => reporter.activityTimer(formatMessage(message)),
    progress: (message, total) => reporter.createProgress(formatMessage(message), total)
  };
}
